Bluetooth: audio: pacs: Move capabilities to PACS, refactor the code

MariuszSkamra opened the pull request:
This moves the code out of capabilities.c and puts it to the pacs.c. Then the pacs.c code have been refactored to cleanup the redundant function calls and to optimize the memory consumption.

This change saves more than 1K flash (taken unicast_audio_server sample as reference)

Before:
```
Memory region         Used Size  Region Size  %age Used
           FLASH:      200712 B         1 MB     19.14%
             RAM:       44679 B       448 KB      9.74%
        IDT_LIST:          0 GB         2 KB      0.00%
```
After:
```
Memory region         Used Size  Region Size  %age Used
           FLASH:      199556 B         1 MB     19.03%
             RAM:       44663 B       448 KB      9.74%
        IDT_LIST:          0 GB         2 KB      0.00%
```

MariuszSkamra committed:
Bluetooth: audio: pacs: Move capabilities to PACS

There is no point of having dedicated source file that is PACS
dependent. This moves all of the code from capabilities.c to pacs.c.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
Bluetooth: audio: pacs: Refactor PAC read handler

This makes use of bt_audio_foreach_capability fuction to
simplify building the PAC attribute value.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
Bluetooth: audio: pacs: Refactor context related code

This implifies and removes redundant code.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
Bluetooth: audio: pacs: Refactor location related code

This simplifies and removes redundant code.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

Thalley commented on subsys/bluetooth/audio/pacs.c:
```diff
@@ -54,124 +81,87 @@ static void pac_data_add(struct net_buf_simple *buf, uint8_t num,
 
-	rsp = net_buf_simple_add(buf, sizeof(*rsp));
-	rsp->num_pac = 0;
+	pac = net_buf_simple_add(buf, sizeof(*pac));
```
We are adding quite a lot of data to the buffer here, but we have no checks for size. This seems dangerous as we do not restrict the number of capabilities we can register. 

If we register 50 capabilities, but can only fit 25 in the buffer (which actually should be `MIN(CONFIG_BT_L2CAP_TX_MTU, BT_ATT_MAX_ATTRIBUTE_LEN)`), should we just put the first 25 in the response, or do something else? 

Alternatively we need to verify the total length of the PAC records when registering capabilities, so that we do no register more than what we can respond with. 

This can be postponed to a different PR, as this is just a move of existing code :)

Thalley commented on subsys/bluetooth/audio/pacs.c:
```diff
@@ -254,38 +234,56 @@ static void snk_cfg_changed(const struct bt_gatt_attr *attr, uint16_t value)
 
+static inline int set_snk_available_contexts(uint16_t contexts)
+{
+	return set_available_contexts(contexts, &snk_available_contexts, snk_supported_contexts);
```
```suggestion
	return set_available_contexts(contexts, &snk_available_contexts, 
				      snk_supported_contexts);
```

Thalley commented on subsys/bluetooth/audio/pacs.c:
```diff
@@ -353,38 +337,56 @@ static void src_cfg_changed(const struct bt_gatt_attr *attr, uint16_t value)
 
+static inline int set_src_available_contexts(uint16_t contexts)
+{
+	return set_available_contexts(contexts, &src_available_contexts, src_supported_contexts);
```
```suggestion
	return set_available_contexts(contexts, &src_available_contexts, 
				      src_supported_contexts);
```

Thalley commented on subsys/bluetooth/audio/pacs.c:
```diff
@@ -585,7 +543,7 @@ static void pac_notify(struct k_work *work)
 }
 
-void bt_pacs_capabilities_changed(enum bt_audio_dir dir)
+static void bt_pacs_capabilities_changed(enum bt_audio_dir dir)
```
Considering removing the `bt_` prefix for static functions

Thalley commented on subsys/bluetooth/audio/pacs.c:
```diff
@@ -597,86 +555,152 @@ void bt_pacs_capabilities_changed(enum bt_audio_dir dir)
 
-int bt_audio_pacs_register_cb(const struct bt_audio_pacs_cb *cb)
+/* Unregister Audio Capability */
+int bt_audio_capability_unregister(enum bt_audio_dir dir, struct bt_audio_capability *cap)
```
A general comment: I agree with the general point of this PR, which is to merge the two files (pacs.c and capabilities.c). However with this direct merge, I think we have some namespace challenges here, as we now have both `bt_pacs` and `bt_audio_capability` in the `pacs.c` file. 

Not sure if this is strictly disallowed in Zephyr, but it certainly isn't particularly great to mix namespaces like that in a single file. 

I'd suggest to use either `bt_audio_capability` or `bt_pacs`, instead of this mix. What do you think?

sjanc replied:
FWIW bt_pacs is much shorter and may suits better as prefix :)

Thalley replied:
> FWIW bt_pacs is much shorter and may suits better as prefix :)

To some extent, yes :) `bt_pacs` does refer to an API for a service (which by now kind of is the case). I am not against that, but it is different than what was originally intended for `bt_audio_capability`, as that was meant to serve as more than just a PACS API I believe (@Vudentz would know, but not sure if he follows this anymore). 

I am OK with using either `bt_audio_capability` or `bt_pacs`. 

Unlike most other functionality, we cannot "hide" the PACS behind a specific role (e.g. `unicast_server` or `broadcast_sink`), as both `unicast_server` and `broadcast_sink` use PACS. 

We will actually have a similar challenge with the client access to this, as PACS can be read by the BAP Broadcast assistant, but all the access to PACS is currently behind the `bt_audio_unicast_client` API. Maybe we will need to consider using `bt_pacs` and `bt_pacs_client` for this purpose (similar to `bt_vocs` and `bt_vocs_client`), where the service may be used by different roles.

MariuszSkamra committed:
Bluetooth: audio: Fix building and reading PAC Records

This fixes missing checks, and invalid struct bt_pac definition that was
missing one mandatory metadata byte.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra commented on subsys/bluetooth/audio/pacs.c:
```diff
@@ -254,38 +234,56 @@ static void snk_cfg_changed(const struct bt_gatt_attr *attr, uint16_t value)
-		BT_DBG("get_pac_loc returned %d", err);
-		return BT_GATT_ERR(BT_ATT_ERR_UNLIKELY);
-	}
+	return bt_gatt_attr_read(conn, attr, buf, len, offset, &location, sizeof(location));
```
ditto, the code limit has been changed some time ago to 100 characters per line

Thalley replied:
Correct, the _maximum_ is 100, but the _recommended_ is 80, and the > 80 limit should only be used if it improves readability significantly. 

I won't block the PR based on these, but the 80 "limit" is still the one we should strive towards. See https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=bdc48fa11e46f867ea4d75fa59ee87a7f48be144

MariuszSkamra committed:
Bluetooth: audio: pacs: Refactor capabilities related code

This simplifies and removes redundant code.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
Bluetooth: audio: pacs: Fix PACS notification

This fixes sending PACS notification with buffer containing invalid,
uninitialized data.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
Bluetooth: audio: pacs: Replace capabilities API with pacs

This changes the API to use proper naming convention, as the code has
been moved out of capabilities.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra commented on subsys/bluetooth/audio/pacs.c:
```diff
@@ -33,9 +33,36 @@
+};
+
+static sys_slist_t snks;
+static sys_slist_t srcs;
```
done

Thalley commented on include/zephyr/bluetooth/audio/pacs.h:
```diff
@@ -17,15 +17,15 @@ extern "C" {
 
 /** @brief Audio Capability structure. */
-struct bt_audio_capability {
+struct bt_pacs {
```
Since are changing the namespace here, we may want to consider Add "Published Audio Capabiltiies" instead of "Audio Capability" in the documentation.

fredrikdanebjer replied:
```suggestion
struct bt_pacs_capabilities {
```

I think even the name should be changed, to me it otherwise signals being a struct for the complete service.

Thalley replied:
@MariuszSkamra There are more cases where we may want to add "Published", e.g. on Like 29-36. Github won't let me comment on those lines, so can't point them out myself :)

MariuszSkamra replied:
done

Thalley commented on include/zephyr/bluetooth/audio/pacs.h:
```diff
@@ -57,7 +57,7 @@ void bt_audio_foreach_capability(enum bt_audio_dir dir, bt_audio_foreach_capabil
  *  @return 0 in case of success or negative value in case of error.
  */
-int bt_audio_capability_register(enum bt_audio_dir dir, struct bt_audio_capability *cap);
+int bt_pacs_register(enum bt_audio_dir dir, struct bt_pacs *cap);
```
```suggestion
int bt_pacs_register(enum bt_audio_dir dir, struct bt_pacs *pac);
```
or maybe

```suggestion
int bt_pacs_register(enum bt_audio_dir dir, struct bt_pacs *pac_record);
```

Thalley commented on samples/bluetooth/hap_ha/src/bap_unicast_sr.c:
```diff
@@ -411,22 +411,22 @@ BT_CONN_CB_DEFINE(conn_callbacks) = {
 };
 
-static struct bt_audio_capability caps_sink = {
+static struct bt_pacs caps_sink = {
```
```suggestion
static struct bt_pacs pac_sink = {
```

Thalley commented on samples/bluetooth/hap_ha/src/bap_unicast_sr.c:
```diff
@@ -411,22 +411,22 @@ BT_CONN_CB_DEFINE(conn_callbacks) = {
 };
 
-static struct bt_audio_capability caps_source = {
+static struct bt_pacs caps_source = {
```
```suggestion
static struct bt_pacs pac_source = {
```

Thalley commented on samples/bluetooth/unicast_audio_server/src/main.c:
```diff
@@ -563,11 +563,11 @@ BT_CONN_CB_DEFINE(conn_callbacks) = {
 };
 
-static struct bt_audio_capability caps_sink = {
+static struct bt_pacs caps_sink = {
```
```suggestion
static struct bt_pacs pac_sink = {
```

Thalley commented on samples/bluetooth/unicast_audio_server/src/main.c:
```diff
@@ -563,11 +563,11 @@ BT_CONN_CB_DEFINE(conn_callbacks) = {
 };
 
-static struct bt_audio_capability caps_source = {
+static struct bt_pacs caps_source = {
```
```suggestion
static struct bt_pacs pac_source = {
```

Thalley commented on subsys/bluetooth/shell/audio.c:
```diff
@@ -684,11 +684,11 @@ static const struct bt_audio_unicast_server_cb unicast_server_cb = {
 };
 
-static struct bt_audio_capability caps_sink = {
+static struct bt_pacs caps_sink = {
```
```suggestion
static struct bt_pacs pac_sink = {
```

Thalley commented on subsys/bluetooth/shell/audio.c:
```diff
@@ -684,11 +684,11 @@ static const struct bt_audio_unicast_server_cb unicast_server_cb = {
 };
 
-static struct bt_audio_capability caps_source = {
+static struct bt_pacs caps_source = {
```
```suggestion
static struct bt_pacs pac_source = {
```

Thalley reviewed (changes requested):
I think this is a significant improvement - Thanks! 

A few comments

MariuszSkamra commented on subsys/bluetooth/audio/pacs.c:
```diff
@@ -33,9 +33,36 @@
+
+#if defined(CONFIG_BT_PAC_SNK)
+static uint16_t snk_available_contexts;
+static const uint16_t snk_supported_contexts = CONFIG_BT_PACS_SNK_CONTEXT;
```
Indeed, current implementation have a Kconfig for supported contexts right now meaning that it cannot be changed.

Thalley replied:
The supported context is something that may change (e.g. if a cable connection is added or removed), so it is something we could add an API for.

fredrikdanebjer commented on include/zephyr/bluetooth/audio/pacs.h:
```diff
@@ -7,96 +7,96 @@
-/** @brief Audio Capability structure. */
-struct bt_audio_capability {
+/** @brief Published Audio Capability structure. */
+struct bt_pacs {
```
```suggestion
struct bt_pacs_capability {
```

I feel the current name more points towards it being a data struct for the entire service.

Thalley replied:
That is a good point. `pacs` is the name of the service, so it may cause some confusion otherwise.

Thalley commented on subsys/bluetooth/audio/pacs.c:
```diff
@@ -33,145 +33,178 @@
-		cc->len = d->data_len + sizeof(cc->type);
-		cc->type = d->type;
+		ltv = net_buf_simple_add(buf, sizeof(*ltv));
+		ltv->len = d->data_len + 1;
```
The addition of `1` here, is due to the type, right?

Perhaps
```suggestion
		ltv->len = d->data_len + sizeof(d->type);
```
or
```suggestion
		ltv->len = d->data_len + sizeof(ltv->type);
```
Would make that clearer

Thalley commented on subsys/bluetooth/audio/pacs_internal.h:
```diff
@@ -22,22 +22,20 @@ struct bt_pac_codec {
-	uint8_t  value[0];		/* Metadata Value */
+struct bt_pac_ltv_hdr {
+	uint8_t  len;
+	uint8_t  data[0];
```
Is there `data` field ever used?

In contrast, it does seem a bit odd to have `data` in a `hdr`

Suggest to remove (and then maybe the header struct is a bit overkill to have).

MariuszSkamra replied:
It's information header for the data behind. It is a common pattern to define network packets. Even if `data[0]` is not explicitly referenced it still gives the information that there is (or not if `len` == 0) data behind.

Thalley replied:
> It is a common pattern to define network packets like this.

Might be, but if we compare it to other `hdr` structs in Zephyr it is very uncommon. In fact, we have 53 structs that define a header struct (based on the `_hdr` suffix), and only 5 of them have a `data` field. 

I am OK with leaving it, but would still suggest to remove (especially since it is also not used).

MariuszSkamra replied:
Ha! Emil, thanks to you, I realized I introduced a bug!

Thalley replied:
> Ha! Emil, thanks to you, I realized I introduced a bug!

I didn't catch that when reviewing :D Is that bug something we can easily add some testing for in the unicast BSIM tests?

MariuszSkamra replied:
It's the matter of size of `struct bt_pac` PAC Record that's minimum size is 7 according to spec.
```
struct bt_pac {
	struct bt_pac_codec codec;	/* Codec ID */
	struct bt_pac_ltv_data cc;	/* Codec Specific Capabilities */
	struct bt_pac_ltv_data meta;	/* Metadata */
} __packed;
```

Thus the `sizeof(struct bt_pac)` has to be 7 bytes :)

Thalley reviewed (changes requested):
Minor comments, otherwise LGTM

Thalley commented on subsys/bluetooth/audio/unicast_client.c:
```diff
@@ -2507,7 +2514,10 @@ static uint8_t unicast_client_read_func(struct bt_conn *conn, uint8_t err,
 		struct unicast_client_pac *bpac;
 		struct bt_pac *pac;
-		struct bt_pac_meta *meta;
+		struct bt_pac_ltv *meta_hdr, *cc_hdr;
```
Here you actually use the `ltv` to store the `hdr`, so perhaps `bt_pac_ltv_hdr` would be a better name? 

It is a bit confusing with the current names where you use the `_ltv` to _not_ store the actual LTV, but rather the header _and_ the LTV.

MariuszSkamra commented:
Rebased
